README.md
# OSS Attribution Builder

OSS Attribution Builder is a website that helps teams create attribution documents for software products. An attribution document is a text file, web page, or screen in just about every software application that lists the software components used and their licenses. They're often in the About screens, and are sometimes labeled "Open Source Notices", "Credits", or other similar jargon.

[Screenshot](https://raw.github.com/amzn/oss-attribution-builder/screenshots/attribution-builder-project-example.png)

## Quickstart

1. Install [Docker](https://www.docker.com/)
2. Clone this repository
3. Run `docker-compose up`
4. Visit http://localhost:8000/
   * The demo uses HTTP basic auth. Enter any username and password. Use `admin` to test out admin functionality.

## Using the Website

See documentation:

* [For users](docs/for-users.md)
* [For administrators](docs/for-admins.md)

## Caveats

The attribution builder was originally an Amazon-internal tool. Some portions had to be removed to make this a sensible open source project. As such, there are some warts:

* Projects have contact lists, but at the moment the UI only supports one contact (the legal contact).

These will all be fixed in time, but be aware that some things might be weird for a while.

## Custom deployment

If you're ready to integrate the attribution builder into your own environment, there are some things to set up:

### Configuration

Open up [config/default.js](config/default.js) and poke around. This configuration launches when you run `docker-compose` or otherwise launch the application.

### Licenses

The attribution builder has support for two types of license definitions:

* SPDX identifiers
* "Known" license texts and tags

SPDX identifiers are just used for pre-filling the license selector, but do not (currently) have texts. The more useful type of license is a "known" license, where **you** (the administrator) supply the text of the license and any tags you'd like to apply.

For information on adding your own "known" licenses, see [the license README](server/licenses/known/README.md). There are two existing licenses in the same directory you can look at for examples.

#### Tags

Tags allow you to add arbitrary validation rules to a license. They can be useful for:

* Verifying a license is being used in the right way (e.g., LGPL and how a package was linked)
* Annotating a particular license as needing follow up, if your business has special processes
* Providing guidance on attribution for licenses with many variants
* Modifying the how a license is displayed in an attribution document

For information on what tags can do and how to create your own, see [the tags README](server/licenses/tags/README.md).

### Extensions

The attribution builder offers some form of extensions that allow you to alter client-side site behavior and appearance, without needing to patch internals. This can make upgrades easier.

See [the extensions README](browser/extensions/README.md) for details.

### Authentication module

The attribution builder supports being able to restrict access to certain people or groups using project ACLs. These can also be used for administration and to "verify" packages (details on that in a later section). The default implementation `nullauth` is not very useful for most environments; you will want to write your own when launching more broadly.

See [the base auth interface](server/auth/base.ts) for implementation details.

### Running

To start up the server, you should run `build/server/localserver.js` after building with `npm run build`. There are some environment variables you'll probably want to set when running:

* `NODE_ENV` should most likely be set to `production`
* `CONFIG_NAME` should be set to the basename (no extension) of your configuration file you created above. The default is "default".

The server runs in HTTP only. You probably want to put a thin HTTPS web server or proxy in front of it.

## Contributing

See [CONTRIBUTING](CONTRIBUTING.md) for information.

### Development

`npm install` and then `npm run dev` will get you off the ground for local development. This will start a Docker container for PostgreSQL, but will use a local copy of tsc, webpack, node, etc so you can iterate quickly.

Once things have started up, you can open http://0.0.0.0:8010/webpack-dev-server/. This will automatically reload on browser changes, and the backend will also automatically restart on server-side changes.

Handy environment variables:

* `NODE_ENV`: when unset or `development`, you'll get full source maps & debug logs
* `DEBUG_SQL`: when set (to anything), this will show SQL queries on the terminal as they execute

#### Testing

`npm test` will run unit tests. These are primarily server focused.

`npm run test-ui` will run Selenium tests. You can set the environment variable `SELENIUM_DRIVER` if you want a custom driver -- by default, it'll try to use Chrome, and if that's not available it'll fall back to PhantomJS.

When debugging UI tests, it may be easier to change `standalone-chrome` to `standalone-chrome-debug` in `docker-compose.selenium.yml`, and then connect to the container via VNC (port 5900, password "secret"). Run the container and your tests separately:

* `docker-compose -f docker-compose.selenium.yml up --build`
* `tsc && jasmine --stop-on-failure=true 'build/selenium/*.spec.js'`

Tests failing for seemingly no reason? `driver.sleep` not working? Make sure your Jasmine timeout on your test is high enough.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html), as much as
a website reasonably can. Backwards incompatible changes (requiring a major version bump) will be
described here, and may involve database changes, significant workflow changes, or changes that
require manual edits to pluggable interfaces.

## Unreleased

### Added
- Added extension points to customize the look and behavior of client-side components. See
  the README in `browser/extensions` for info.
- A global ACL is now available and can be set in your configuration as `globalACL`. Entries added
  to this list will implicitly apply to all projects. This will eventually replace the current
  admin functionality.
- An API is available to fetch rendered/stored documents for projects.

### Changed
- Now uses tiny-attribution-generator to build documents.

## 0.9.0 - 2017-12-11

### Added
- Auth backends can now specify how a user should be authenticated, via Passport. They should
  provide an `initialize` method that is called during app start-up. This can be used to register
  Passport strategies, login URLs, or any other session activities.
- SPDX license texts are now shipped with the attribution builder.
- License tags can now specify presentation options to influence how they appear in the package
  editor. They can be sorted first, annotated with text (both in menu and below), and control
  whether users are asked for the full license text.
- License tags can also specify "questions" to ask a user when adding a package. This is useful
  to gather context-sensitive info. For example, you could only ask for "dynamic/static linking"
  if relevant for a given license.
- Added a user interface for editing project access lists. This can be accessed by clicking on
  the owner on the top right side of the projcet editor.
- It is now possible to edit a package and usage information in a project. New package revisions
  will be created as necessary, and previous entries will be correctly cleaned up.

### Removed
- JWT sessions are no longer in use. See the above addition about auth backends for an alternative.
- The build process no longer requires Gulp.

### Changed
- Project ACLs are now sanely validated, with levels of "owner", "editor", and "viewer". A viewer
  can only view a project. An editor can change project details, except for the ACL. An owner can
  change everything about a project.
- Users on a project contact list implicitly have "viewer" permissions unless otherwire specified.
- The format of `/api/licenses` changed. Instead of a list, it returns a {licenses, tags}
  structure. The license list is included in the `license` key.

### Fixed
- Some lingering Bootstrap CSS issues were cleaned up.
- The `validateUsage` function (used in tags) was incorrectly documented.
- `extractRequestUser` is now consistently used, making custom auth backends more reliable.

### Security
- Users who weren't configured to access package validation systems could still do so, due to
  a dangling `Promise`. Additional type checks and lints have been enabled to prevent this in the
  future.

## 0.8.0 - 2017-08-04

- Initial release.

gulpfile.js
/* Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
*
* Licensed under the Apache License, Version 2.0 (the "License").
* You may not use this file except in compliance with the License.
* A copy of the License is located at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* or in the "license" file accompanying this file. This file is distributed
* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
* express or implied. See the License for the specific language governing
* permissions and limitations under the License.
*/

// @ts-nocheck
'use strict';

const spawn = require('child_process').spawn;
const gulp = require('gulp');
const jasmine = require('gulp-jasmine');
const plumber = require('gulp-plumber');
const sourcemaps = require('gulp-sourcemaps');
const ts = require('gulp-typescript');
const webpack = require('webpack');
const webpackStream = require('webpack-stream');
const WebpackDevServer = require('webpack-dev-server');

const tsProject = ts.createProject('tsconfig.json');
gulp.task('build-server', ['copy-config', 'copy-data'], () => {
  return gulp.src(['server/**/*.ts'])
    .pipe(plumber())
    .pipe(sourcemaps.init())
    .pipe(tsProject())
    .pipe(sourcemaps.write('.'))
    .pipe(gulp.dest('build/server'));
});

gulp.task('build-selenium', () => {
  return gulp.src(['selenium/**/*.ts'])
    .pipe(plumber())
    .pipe(sourcemaps.init())
    .pipe(tsProject())
    .pipe(sourcemaps.write('.'))
    .pipe(gulp.dest('build/selenium'));
});

gulp.task('build-browser', () => {
  return gulp.src('browser/app.tsx')
    .pipe(plumber())
    .pipe(webpackStream(require('./webpack.config.js'), webpack))
    .pipe(gulp.dest('build/res'));
});

gulp.task('copy-assets', () => {
  return gulp.src('assets/**/*')
    .pipe(gulp.dest('build/server/assets'));
});

gulp.task('copy-config', () => {
  return gulp.src('config/**/*')
    .pipe(gulp.dest('build/config'));
});

gulp.task('copy-data', () => {
  return gulp.src('data/**/*.json')
    .pipe(gulp.dest('build/data'));
});

gulp.task('test', ['build-server'], () => {
  require('source-map-support/register');
  require('core-js/shim');
  const winston = require('winston');
  winston.remove(winston.transports.Console);
  process.on('unhandledRejection', err => {console.error('\nUnhandled rejection:', err.stack)});
  return gulp.src('build/server/**/*.spec.js')
    .pipe(jasmine());
});

let node;
gulp.task('local-server', ['build-server'], done => {
  if (node) {
    node.kill();
  }

  // set development mode if unset
  process.env.NODE_ENV = process.env.NODE_ENV == null ? 'development' : process.env.NODE_ENV;

  // launch it
  node = spawn('node', ['build/server/localserver.js']);

  // pass through stdout and wait on a "ready" message in output
  node.stdout.on('data', data => {
    const text = data.toString();
    process.stdout.write(text);
    if (/server running/i.test(text)) {
      done();
    }
  });

  // pass through stderr
  node.stderr.on('data', data => {
    process.stderr.write(data.toString());
  });

  // if it exits early (i.e., not killed for reload) then finish with error
  node.on('close', (code, signal) => {
    if (signal == null) {
      done(`server exited prematurely (code ${code})`);
    }
  });
});

gulp.task('dev', ['copy-assets', 'local-server'], () => {
  gulp.watch(['server/**/*.ts'], {debounceDelay: 500}, ['local-server']);
  gulp.watch(['assets/**/*'], ['copy-assets']);

  const compiler = webpack(require('./webpack.config.js'));
  new WebpackDevServer(compiler, {
    publicPath: '/res/',
    proxy: {'*': 'http://localhost:8000'},
    stats: 'minimal',
  }).listen(8010, '0.0.0.0', err => {
    if (err) {
      throw new Error('webpack-dev-server error:', err);
    }
    console.log('Autoreload server running at http://0.0.0.0:8010/webpack-dev-server/');
  });
});

gulp.task('default', ['build-server', 'build-browser', 'copy-assets']);

process.on('exit', () => {
  if (node) {
    node.kill();
  }
});
